src/algorithms/lca.rs

```diff
@@ -0,0 +1,287 @@
+                                parent: Some(node),
+                            });
+                        }
+                    }
+                    DFSState::Finish { node } => {
+                        lca.last_visit[node] = timestamp;
+                        timestamp += 1;
+                    }
+                }
+            }
+        }
+
+        lca
+    }
+
+    /// Returns `true` if `a` is an ancestor of `b` in the hierarchy.
+    ///
+    /// If `a` and `b` are the same node, returns `true`.
+    ///
+    /// Complexity: O(1)
+    pub fn is_ancestor(&self, a: NodeIndex, b: NodeIndex) -> bool {
+        self.first_visit[a] <= self.first_visit[b] && self.last_visit[a] >= self.last_visit[b]
+    }
+
+    /// Returns the root of the hierarchy that contains the given node.
+    ///
+    /// Complexity: O(log n)
+    pub fn root(&self, node: NodeIndex) -> NodeIndex {
+        let mut u = node;
+        while let Some(&v) = self.climb_nodes[u].last() {
+            u = v;
+        }
+        u
+    }
+
+    /// Given two nodes, returns the lowest common ancestor in the hierarchy.
+    ///
+    /// If the nodes are not in the same tree, returns `None`.
+    ///
+    /// Complexity: O(log n)
+    pub fn lca(&self, a: NodeIndex, b: NodeIndex) -> Option<NodeIndex> {
+        if self.is_ancestor(a, b) {
+            return Some(a);
+        }
+        if self.is_ancestor(b, a) {
+            return Some(b);
+        }
+        // The nodes are in different trees.
+        if self.root(a) != self.root(b) {
+            return None;
+        }
+
+        // Search the ancestors of `a` to find the lowest common ancestor with `b`.
+        //
+        // Invariant: `u` is an ancestor of `a` (or `a`), but not an ancestor of `b`.
+        //
+        // We start by searching a `u` where the last ancestor in the climb nodes
+        // is an ancestor of `b`.
+        let mut u = a;
+        loop {
+            let Some(&last_climb) = self.climb_nodes[u].last() else {
+                // We reached a root, and it is not an ancestor of `b`.
+                return None;
+            };
+            if self.is_ancestor(last_climb, b) {
+                // We found a `u` where the last ancestor is an ancestor of `b`.
+                break;
+            }
+            u = last_climb;
+        }
+
+        // Invariant: The 2^i ancestor of `u` is an ancestor of `b`.
+        //
+        // We search the ancestors of a, each time decreasing `i`.
+        let mut i = self.climb_nodes[u].len() - 1;
+        while i > 0 {
+            i -= 1;
+            // The 2^i ancestor of `u`.
+            let v = self.climb_nodes[u][i];
+            if !self.is_ancestor(v, b) {
+                // The 2^i ancestor is not an ancestor of `b`.
+                // Lets update `u`.
+                u = v;
+                // Decrease `i`, and ensure it is within bounds.
+                i = i.max(self.climb_nodes[u].len() - 1);
```

Comment:
I've added some comments that made it clearer for me. You could add another comment that this is binary search and that `2^i` is the size of the interval the LCA must be in.
```suggestion
                // The 2^i ancestor of `u` is not an ancestor of `b` so the LCA must be between
                // the 2^i and 2^{i+1} ancestor of `u`. Hence update `u` to the 2^i ancestor.
                u = v;
                // Ensure `i` is within bounds.
                i = i.max(self.climb_nodes[u].len() - 1);
```

Reply:
👍 
I added more comments.